keps/sig-cluster-lifecycle/kubeadm/2568-kubeadm-non-root-control-plane/README.md

```diff
@@ -0,0 +1,1069 @@
+
+3. Set the seccomp profile to `runtime/default`, this can be done by setting the `seccompProfile` field in the `securityContext`.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a security conscious user I would like to run the kubernetes control-plane as non-root.
+
+#### Story 2
+
+As a `kubeadm` user, I'd expect the bootstrapper to follow the best security practices when generating static or non-static pod manifests to run control plane components.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+There are some caveats with running the components as non-root around how to assign `UID`/`GID`s to them. This is covered in the [Assigning UID and GID](#assigning-uid-and-gid) section below. 
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+If we hard coded the `UID` and `GID`, we could end up in a scenario where those are in use by another process on the machine, which would expose some of the credentials accessible to the `UID` and `GID`s to that process. So we plan to use adduser --system or using the appropriate ranges from /etc/login.defs instead of hard coding the `UID` and `GID`.
+
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+In `kubeadm` the control-plane components are run as [static-pods](https://kubernetes.io/docs/concepts/workloads/pods/#static-pods), i.e. pods directly managed by kubelet. We can use the `runAsUser` and `runAsGroup` fields in [SecurityContext](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#securitycontext-v1-core) to run the containers in these Pods with a unique `UID` and `GID` and the `capabilities` field to drop all capabilities other than the ones required.
+
+### Assigning UID and GID
+There are 3 options for setting the `UID`/`GID` of the control-plane components:-
```

Comment:
Another option here assuming we will be naming the UID and GID the same as the component they own, is to generate them in such a way that UID == GID == Process ID when running these components as non-root the first time. This is not a perfect solution and has following pros and cons:

Pros:
1. Allows mapping of UID and GID to components via Process ID. This might be helpful and intuitive during production troubleshooting
2. Ensures uniqueness at a point in time as no two processes at any given point in time will have same process ID.
3. Most likely will never take UID = 0 as the process ID for each of the components being zero is highly unlikely, since something (e.g. swapper or scheduler) might be running with Process ID == 0

Cons:
1. Might need some statefulness to pick up the the same UID and GID next time any of these components are started or restarted as the process ID could be different 
2. Might have some overlap with existing UID or Usernames e.g. process ID of api-server is 1001. A user with UID 1001 exists and is running a monitoring service in the same node as api-server. This could lead to some interesting privilege escalation attacks.

Reply:
We could also ensure uniqueness if we used adduser and groupadd.
This would add complications on restarts. restarts which are completely managed by kubelet today would now need to be managed by kubeadm somehow to update not only the manifest with the new uid but also all file permissions, this would be fairly complicated.

Reply:
That makes sense. If it is not too much trouble, could you add this as 4th option in KEP markdown and share why this is not something that was chosen. I worry that these comments will get lost after KEP is merged and we may end up revisiting this in future. Putting this in KEP will save some future time for all of us :) Happy to make edits to the KEP myself but I am afraid I do not have permissions to do the same.

Reply:
Happy to add it as an option. Can you suggest it as an edit?

Reply:
We should say the reasoning for following PAM SYS_UID_MIN is to fall in line with distro standards. I did link the relevant docs to Debian and Fedora packaging in a previous comment.

Reply:
Good call, done!